framework/src/modules/interoperability/utils.ts

```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
+		MAINCHAIN_ID_BUFFER,
+		ownChainAccountSchema,
+	);
+	const channelDataStoreKeySet = new dataStructures.BufferSet();
+	const channelDataStore = getStore(id, STORE_PREFIX_CHANNEL_DATA);
+	for (const channelData of genesisStore.channelDataSubstore) {
+		if (channelDataStoreKeySet.has(channelData.storeKey)) {
+			throw new Error(
+				`Channel data store key ${channelData.storeKey.toString('hex')} is duplicated.`,
+			);
+		}
+		channelDataStoreKeySet.add(channelData.storeKey);
+
+		const tokenID = channelData.storeValue.messageFeeTokenID;
+		if (
+			!(
+				tokenID.chainID.readInt32BE(0) === 1 ||
+				tokenID.chainID.equals(channelData.storeKey) ||
+				tokenID.chainID.equals(ownChainAccount.id)
+			)
+		) {
+			throw new Error(
+				`Chain id corresponding to the channel data store key ${channelData.storeKey.toString(
+					'hex',
+				)} is not valid.`,
+			);
+		}
+		if (tokenID.chainID.equals(MAINCHAIN_ID_BUFFER) && tokenID.localID.readInt32BE(0) !== 0) {
+			throw new Error(
+				`Local id corresponding to the channel data store key ${channelData.storeKey.toString(
+					'hex',
+				)} is not valid.`,
+			);
+		}
+
+		await channelDataStore.setWithSchema(
+			channelData.storeKey,
+			channelData.storeValue,
+			channelSchema,
+		);
+	}
+
+	const chainValidatorsStoreKeySet = new dataStructures.BufferSet();
+	const chainValidatorsStore = getStore(id, STORE_PREFIX_CHAIN_VALIDATORS);
+	for (const chainValidators of genesisStore.chainValidatorsSubstore) {
+		if (chainValidatorsStoreKeySet.has(chainValidators.storeKey)) {
+			throw new Error(
+				`Chain validators store key ${chainValidators.storeKey.toString('hex')} is duplicated.`,
+			);
+		}
+		chainValidatorsStoreKeySet.add(chainValidators.storeKey);
+
+		const { activeValidators, certificateThreshold } = chainValidators.storeValue;
+		if (activeValidators.length < 1 || activeValidators.length > MAX_NUM_VALIDATORS) {
+			throw new Error(
+				`Active validators must have at least 1 element and at most ${MAX_NUM_VALIDATORS} elements.`,
+			);
+		}
+
+		let totalWeight = BigInt(0);
+		for (let j = 0; j < activeValidators.length; j += 1) {
+			const activeValidator = activeValidators[j];
+
+			const { blsKey } = activeValidator;
+			if (j < activeValidators.length - 1 && blsKey.compare(activeValidators[j + 1].blsKey) >= 0) {
+				throw new Error(
+					'Active validators must be ordered lexicographically by blsKey property and pairwise distinct.',
+				);
+			}
+
+			const { bftWeight } = activeValidator;
+			totalWeight += bftWeight;
+		}
+
+		if (totalWeight > MAX_UINT64) {
+			throw new Error(
+				'The total BFT weight of all active validators has to be less than or equal to MAX_UINT64.',
+			);
+		}
+
+		const checkBftWeightValue = totalWeight / BigInt(3) + BigInt(1);
+		if (checkBftWeightValue > totalWeight || checkBftWeightValue > certificateThreshold) {
+			throw new Error('The total BFT weight of all active validators is not valid.');
+		}
```

Comment:
Check that ⌊ ⅓ x totalWeight⌋ + 1 <= certificateThreshold <= totalWeight, where ⌊⋅⌋ is the floor function.